File: app/src/main/java/com/qimian233/ztool/LoadingDialog.java
package com.qimian233.ztool;

import android.app.Activity;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import androidx.appcompat.app.AlertDialog;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

public class LoadingDialog {
    private AlertDialog dialog;
    private TextView tvDescription;
    private ProgressBar progressBar;
    private final Context context;
    private boolean isShowing = false;

    public LoadingDialog(Context context) {
        this.context = context;
        initDialog();
    }

    private void initDialog() {
        // 使用MaterialAlertDialogBuilder创建对话框
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(context);

        // 加载自定义布局
        LayoutInflater inflater = LayoutInflater.from(context);
        View dialogView = inflater.inflate(R.layout.dialog_loading, null);

        // 初始化视图
        progressBar = dialogView.findViewById(R.id.progress_bar);
        tvDescription = dialogView.findViewById(R.id.tv_description);

        builder.setView(dialogView);

        // 设置对话框不可关闭
        dialog = builder.create();
        dialog.setCancelable(false);
        dialog.setCanceledOnTouchOutside(false);
    }

    /**
     * 显示加载对话框
     * @param message 初始描述文本
     */
    public void show(String message) {
        if (dialog != null && !isShowing) {
            tvDescription.setText(message);
            dialog.show();
            isShowing = true;
        }
    }

    /**
     * 显示加载对话框（使用默认文本）
     */
    public void show() {
        show("加载中...");
    }

    /**
     * 隐藏对话框
     */
    public void dismiss() {
        if (dialog != null && isShowing) {
            // 确保在主线程中执行
            if (Looper.myLooper() == Looper.getMainLooper()) {
                dialog.dismiss();
            } else {
                new Handler(Looper.getMainLooper()).post(() -> dialog.dismiss());
            }
            isShowing = false;
        }
    }

    /**
     * 动态更新描述文本
     * @param message 新的描述文本
     */
    public void updateMessage(String message) {
        if (tvDescription != null && isShowing) {
            if (Looper.myLooper() == Looper.getMainLooper()) {
                tvDescription.setText(message);
            } else {
                new Handler(Looper.getMainLooper()).post(() -> tvDescription.setText(message));
            }
        }
    }

    /**
     * 检查对话框是否正在显示
     */
    public boolean isShowing() {
        return isShowing;
    }

    /**
     * 设置对话框是否可取消
     */
    public void setCancelable(boolean cancelable) {
        if (dialog != null) {
            dialog.setCancelable(cancelable);
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/LoadingDialog.kt
package com.qimian233.ztool

import android.app.Dialog
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.ViewGroup
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.qimian233.ztool.ui.components.ZToolCircularProgressIndicator
import com.qimian233.ztool.ui.components.createPlatformComposeDialog

class LoadingDialog(private val context: Context) {
    private val handler = Handler(Looper.getMainLooper())
    private var dialog: Dialog? = null
    private var showing = false
    private var message by mutableStateOf(context.getString(R.string.loading))

    fun show(message: String) {
        runOnMain {
            this.message = message
            if (!showing) {
                ensureDialog().show()
                ensureDialog().window?.apply {
                    setLayout(
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT
                    )
                    setGravity(Gravity.CENTER)
                }
                showing = true
            }
        }
    }

    fun show() {
        show(context.getString(R.string.loading))
    }

    fun dismiss() {
        runOnMain {
            if (showing) {
                dialog?.dismiss()
                showing = false
            }
            handler.removeCallbacksAndMessages(null)
            dialog = null
        }
    }

    @Suppress("unused")
    fun updateMessage(message: String) {
        runOnMain {
            if (showing) {
                this.message = message
            }
        }
    }

    @Suppress("unused")
    fun isShowing(): Boolean = showing

    private fun ensureDialog(): Dialog {
        val existing = dialog
        if (existing != null) return existing

        return createPlatformComposeDialog(context, cancelable = false) {
            Surface(color = MaterialTheme.colorScheme.surface) {
                Row(
                    modifier = Modifier.padding(24.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    ZToolCircularProgressIndicator(modifier = Modifier.size(40.dp))
                    Spacer(modifier = Modifier.width(16.dp))
                    Text(
                        text = message,
                        style = MaterialTheme.typography.bodyLarge,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.fillMaxWidth()
                    )
                }
            }
        }.also {
            dialog = it
        }
    }

    private fun runOnMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            block()
        } else {
            handler.post(block)
        }
    }
}
